pkg/scheduler/queue/tenant_queues.go

```diff
@@ -190,17 +193,39 @@ func (qb *queueBroker) dequeueRequestForQuerier(
 // but there is no reason to make consumers know that they need to call through to the tenantQuerierAssignments.
 
 func (qb *queueBroker) addQuerierWorkerConn(conn *QuerierWorkerConn) (resharded bool) {
-	return qb.tenantQuerierAssignments.addQuerierWorkerConn(conn)
+	// if conn is for a new querier, we need to recompute tenant querier relationship; otherwise, we don't reshard
+	if addQuerier := qb.querierConnections.addQuerierWorkerConn(conn); addQuerier {
+		qb.tenantQuerierAssignments.addQuerier(conn.QuerierID)
+		return qb.tenantQuerierAssignments.recomputeTenantQueriers()
+	}
+	return false
+	//return qb.tenantQuerierAssignments.addQuerierWorkerConn(conn)
 }
 
 func (qb *queueBroker) removeQuerierWorkerConn(conn *QuerierWorkerConn, now time.Time) (resharded bool) {
-	return qb.tenantQuerierAssignments.removeQuerierWorkerConn(conn, now)
+	// if we're removing the last active connection for the querier, the querier may need to be removed
+	if removedQuerier := qb.querierConnections.removeQuerierWorkerConn(conn, now); removedQuerier {
+		return qb.tenantQuerierAssignments.removeQueriers(conn.QuerierID)
+	}
+	return false
 }
 
+// notifyQuerierShutdown handles a graceful shutdown notification from a querier.
+// Returns true if tenant-querier reshard was triggered.
 func (qb *queueBroker) notifyQuerierShutdown(querierID QuerierID) (resharded bool) {
-	return qb.tenantQuerierAssignments.notifyQuerierShutdown(querierID)
+	if removedQuerier := qb.querierConnections.shutdownQuerier(querierID); removedQuerier {
+		qb.tenantQuerierAssignments.removeQueriers(querierID)
+		return qb.tenantQuerierAssignments.recomputeTenantQueriers()
+	}
+	return false
 }
 
+// forgetDisconnectedQueriers removes all queriers which have had zero connections for longer than the forget delay.
+// Returns true if tenant-querier reshard was triggered.
 func (qb *queueBroker) forgetDisconnectedQueriers(now time.Time) (resharded bool) {
-	return qb.tenantQuerierAssignments.forgetDisconnectedQueriers(now)
+	queriersToRemove := qb.querierConnections.forgettableQueriers(now)
+	for _, querierID := range queriersToRemove {
+		qb.querierConnections.removeQuerier(querierID)
+	}
+	return qb.tenantQuerierAssignments.removeQueriers(queriersToRemove...)
```

Comment:
IMO we might have decomposed the steps a touch too much here - if we have no usage of `forgettableQueriers` where we do anything but remove each one, I think we should wrap it up into one interface:

instead of this
```golang
	queriersToRemove := qb.querierConnections.forgettableQueriers(now)
	for _, querierID := range queriersToRemove {
		qb.querierConnections.removeQuerier(querierID)
	}
	return qb.tenantQuerierAssignments.removeQueriers(queriersToRemove...)
```

we could have

```golang
forgottenQueriers := qb.querierConnections.removeForgettableQueriers(now)
return qb.tenantQuerierAssignments.removeQueriers(forgottenQueriers...)
```

So `queueBroker.forgetDisconnectedQueriers` is still serving this responsibility of "coordinating the two things that need done at the same time", but the `querierConnections` is still encapsulating the full step it is responsible for instead of making the caller do it